### PR TITLE
LSP: completion capability

### PR DESCRIPTION
**Pull request by joacoc**
### Motivation

This PR implements an initial completion feature for the LSP server. To use this feature, the client needs to send the schema to the server, and the server will build the completion items. The server only builds completion items for two keywords (`SELECT` and `FROM`.) The completion items returned for these keywords are: 

* `SELECT`: Returns all the available columns in the schema.
* `FROM`: Returns all the available objects (tables, views, materialized views, sources and sinks) available in the schema.

If the keyword is not one of those mentioned, the server returns nothing to avoid confusion. Functions and snippets were not included in this PR to make it simpler and smaller.

<details open>
<summary>This is how it looks in VS Code</summary>
Screenshot 2023-11-24 at 19 31 21
Screenshot 2023-11-24 at 19 30 43
</details>

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/lsp-server/src/backend.rs [hunk 1]
```diff
@@ -91,37 +105,118 @@ pub struct Backend {
...
+}
+
+/// Represents a column from an [ObjectType
+#[derive(Debug, Clone, Deserialize)]
+pub struct SchemaObjectColumn {
+    /// Represents the column's name.
+    pub name: String,
+    /// Represents the column's type.
+    #[serde(rename = "type")]
+    pub _type: String,
```
Comment (mjibson) [inline src/lsp-server/src/backend.rs:130]: `_` causes things to be ignored in rust. I recommend using `typ` as the name instead to get those benefits.

File: src/lsp-server/src/backend.rs [hunk 2]
```diff
@@ -91,37 +105,118 @@ pub struct Backend {
...
+    View,
+    /// Represents a source.
+    Source,
+    /// Represents a table.
+    Table,
+    /// Represents a sink.
+    Sink,
+}
+
+impl ToString for ObjectType {
```
Comment (mjibson) [inline src/lsp-server/src/backend.rs:149]: Generally we would implement `std::fmt::Display` instead, and you get to string for free.

File: src/lsp-server/src/backend.rs [hunk 3]
```diff
@@ -297,6 +416,62 @@ impl LanguageServer for Backend {
         Ok(None)
     }
 
+    /// Completion implementation.
+    async fn completion(&self, params: CompletionParams) -> Result<Option<CompletionResponse>> {
+        let _uri = params.text_document_position.text_document.uri;
```
Comment (mjibson) [inline src/lsp-server/src/backend.rs:421]: These should not have _

File: src/lsp-server/src/backend.rs [hunk 4]
```diff
@@ -297,6 +416,62 @@ impl LanguageServer for Backend {
...
+                .map_err(|_| build_error("Error getting lex tokens."))?;
+            let offset = position_to_offset(_position, content)
+                .ok_or(build_error("Error getting completion offset."))?;
+            let mut last_keyword: Option<Keyword> = None;
+            lex_results.iter().find(|x| {
+                if x.offset < offset {
+                    match x.kind {
+                        Token::Keyword(k) => match k {
+                            Keyword::Select => {
+                                last_keyword = Some(k);
```
Comment (mjibson) [inline src/lsp-server/src/backend.rs:439]: You might need `.filter_map(..).last()` here instead. Using `find` like and mutating external state is not common practice.

**Comment by joacoc** [comment]
Thanks, @mjibson!

**Review by mjibson — APPROVED** [review]

**Pull request merged**